src/Ouroboros/Network/MonadClass/MonadSTM.hs

```diff
@@ -184,3 +204,57 @@ isEmptyTMVarDefault (TMVar t) = do
     Nothing -> return True
     Just _  -> return False
 
+data TBQueueDefault m a = TBQueue
+  !(TVar m Natural) -- read capacity
+  !(TVar m [a])     -- elements waiting for read
+  !(TVar m Natural) -- write capacity
+  !(TVar m [a])     -- written elements
+  !Natural
+
+newTBQueueDefault :: MonadSTM m => Natural -> Tr m (TBQueueDefault m a)
+newTBQueueDefault size = do
+  rsize <- newTVar 0
+  read  <- newTVar []
+  wsize <- newTVar size
+  write <- newTVar []
+  return (TBQueue rsize read wsize write size)
+
+readTBQueueDefault :: MonadSTM m => TBQueueDefault m a -> Tr m a
+readTBQueueDefault (TBQueue rsize read _wsize write _size) = do
+  xs <- readTVar read
+  r <- readTVar rsize
+  writeTVar rsize $! r + 1
+  case xs of
+    (x:xs') -> do
+      writeTVar read xs'
+      return x
+    [] -> do
+      ys <- readTVar write
+      case ys of
+        [] -> retry
+        _  -> do
+          let (z:zs) = reverse ys -- NB. lazy: we want the transaction to be
+                                  -- short, otherwise it will conflict
+          writeTVar write []
+          writeTVar read zs
+          return z
+
+writeTBQueueDefault :: MonadSTM m => TBQueueDefault m a -> a -> Tr m ()
+writeTBQueueDefault (TBQueue rsize _read wsize write _size) a = do
+  w <- readTVar wsize
+  if (w > 0)
+    then do writeTVar wsize $! w - 1
+    else do
+          r <- readTVar rsize
+          if (r > 0)
+            then do writeTVar rsize 0
+                    writeTVar wsize $! r - 1
+            else retry
+  listend <- readTVar write
+  writeTVar write (a:listend)
+
+lengthTBQueueDefault :: MonadSTM m => TBQueueDefault m a -> Tr m Natural
+lengthTBQueueDefault (TBQueue rsize _read wsize _write size) = do
+  r <- readTVar rsize
+  w <- readTVar wsize
+  return $! size - r - w
```

Comment:
I'm need to point out again my objection to writing a clone of `TBQueue` (or any other existing thing in this way). I understand that the motivation is to use a `TBQueue` in simulation, but I don't believe that's a good way to about things.

The shallow/obvious objection is that we're copying an existing library implementation, and so we have to keep tracking it for updates (if the true `stm` queue gets a bug fix, then our simulation is not faithful until it gets the fix).

But prior to that, the scope is just wrong. We want to simulate a _queue_, not a ball of `TVar`s. When we run a simulation I think we'll find it rather useless to get a trace of `TVar` writes and transaction retries; we want something higher level i.e. _queue_ read/writes and other _queue_-related things. That's all to say: we should be abstracting over a _queue_, not writing a single queue over an abstract `STM`.

Reply:
This brings back us back to the discussion on simulation.  Let's continue this discussion on slack https://input-output-rnd.slack.com/archives/CDA6LUXAQ/p1542965749348100 where it will be exposed to more developers.